v2.0/performance-tuning.md

```diff
@@ -0,0 +1,1275 @@
+| NULL                                                                     | /"boston"/"v_\x8a\xf3\x0emJ\x82\x95nb\u0093\x0e\x97q"/PrefixEnd          |       52 | {1,2,3}  |            2 |
+| /"boston"/"v_\x8a\xf3\x0emJ\x82\x95nb\u0093\x0e\x97q"/PrefixEnd          | /"los                                                                    |       42 | {1,2,3}  |            3 |
+|                                                                          | angeles"/"\xe9\xb3\u007f\xbd\x01\x9dLT\xa9~\x9dc\x0f\xa1\x8a-"/PrefixEnd |          |          |              |
+| /"los                                                                    | /"paris"/"b\xa7\x01\xf9\xd6\x0fN\xbc\x81\xb5:\\\x963l="/PrefixEnd        |       46 | {1,2,3}  |            2 |
+| angeles"/"\xe9\xb3\u007f\xbd\x01\x9dLT\xa9~\x9dc\x0f\xa1\x8a-"/PrefixEnd |                                                                          |          |          |              |
+| /"paris"/"b\xa7\x01\xf9\xd6\x0fN\xbc\x81\xb5:\\\x963l="/PrefixEnd        | /"rome"/"\xe7:\xe2 {\x1bE\x81\x97E\x1c\xe0XG\x84\xfb"/PrefixEnd          |       47 | {1,2,3}  |            2 |
+| /"rome"/"\xe7:\xe2 {\x1bE\x81\x97E\x1c\xe0XG\x84\xfb"/PrefixEnd          | /"san francisco"/"\xb0\xa2\x1c\x87!kDx\xa8\x857n\v\a\xdcA"               |       49 | {1,2,3}  |            2 |
+| /"san francisco"/"\xb0\xa2\x1c\x87!kDx\xa8\x857n\v\a\xdcA"               | /"washington dc"/"$}Y\xb0\x15e@㩙\x8b\xe4U\x94\xf0q"/PrefixEnd           |       51 | {1,2,3}  |            2 |
+| /"washington dc"/"$}Y\xb0\x15e@㩙\x8b\xe4U\x94\xf0q"/PrefixEnd           | NULL                                                                     |       41 | {1,2,3}  |            2 |
++--------------------------------------------------------------------------+--------------------------------------------------------------------------+----------+----------+--------------+
+(7 rows)
+~~~
+
+{% include copy-clipboard.html %}
+~~~ shell
+$ cockroach sql \
+--insecure \
+--host=<address of any node> \
+--database=movr \
+--execute="SHOW EXPERIMENTAL_RANGES FROM TABLE users;"
+~~~
+
+~~~
++-----------+---------+----------+----------+--------------+
+| Start Key | End Key | Range ID | Replicas | Lease Holder |
++-----------+---------+----------+----------+--------------+
+| NULL      | NULL    |       27 | {1,2,3}  |            1 |
++-----------+---------+----------+----------+--------------+
+(1 row)
+~~~
+
+The results above tell us that the `rides` table is split across 7 ranges, with 1 range's leaseholder on node 3 and the leaseholders for the rest of the ranges on node 2, and the leaseholder of the single range for `users` is on node 1.
+
+With this information, we can visualize what's happening, assuming the request is sent to node 1 and ignoring non-involved ranges:
+
+<img src="{{ 'images/v2.0/perf_tuning_join1.png' | relative_url }}" alt="Perf tuning concepts" style="max-width:100%" />
```

Comment:
The previous diagrams were OK to follow along; this one in contrast has me thoroughly confused: I do not understand the meaning of the arrows that jump from one "lease" box to another on node 2.

If you mean to explain that node 1 is forwarding the full table scan to every range in the "rides" table, that must be done by using:

- 6 arrows from node 1 to node 2 (you may group them together because they actually flow along a single network link)
- 1 arrow from node 1 to node 3 
- 6 arrows back from node 2 to the outside of node 1 (for the results)
- 1 arrow back from node 3 to the outside of node 1

Then for the join there is no extra arrow needed because node 1 is already leaseholder.


There should not be any arrow from node 2 to itself, and neither from node 2 to the "lease" box of node 1.

Reply:
Thanks, @knz. Do you think this section would be sufficient without the diagram? I think they were important in the intro concepts, but I'm unsure whether they're as useful embedded in the tutorial.

Reply:
I think the diagram is not necessary here indeed.

Reply:
Removed it.